model_zoo/moe/gpt.py

```diff
@@ -0,0 +1,217 @@
+            else:
+                num_experts = num_experts_list[i // 2]
+                block_module = MOEGPTBlock(num_experts=num_experts,
+                                           capacity_factor_train=capacity_factor_train,
+                                           capacity_factor_eval=capacity_factor_eval,
+                                           use_residual=use_residual,
+                                           **block_factory)
+
+            block_list.append(block_module)
+
+        self.blocks = nn.ModuleList(block_list)
+
+        self.norm = col_nn.LayerNorm(normalized_shape=dim, eps=layernorm_epsilon, dtype=dtype)
+
+        self.head = GPTLMHead(dim=dim,
+                              vocab_size=vocab_size,
+                              word_embeeding_weight=self.embed.word_embedding_weight,
+                              dtype=dtype)
+
+    def forward(self, input_ids, attention_mask=None):
+        MOE_CONTEXT.reset_loss()
+        x = self.embed(input_ids)
+
+        # We create a 3D attention mask from a 2D tensor mask.
+        # Sizes are [batch_size, 1, 1, to_seq_length]
+        # So we can broadcast to [batch_size, num_heads, from_seq_length, to_seq_length]
+        # Adapted from huggingface
+        if attention_mask is not None:
+            batch_size = input_ids.shape[0]
+            attention_mask = attention_mask.view(batch_size, -1)
+            attention_mask = col_nn.partition_batch(attention_mask)
+            attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
+            attention_mask = attention_mask.to(dtype=x.dtype)    # fp16 compatibility
+            attention_mask = (1.0 - attention_mask) * -10000.0
+
+        for block in self.blocks:
+            x, attention_mask = block(x, attention_mask)
+
+        x = self.head(self.norm(x))
+
+        return x
+
+
+def _create_moegpt_model(**model_kwargs):
+    model = MOEGPT(**model_kwargs)
+    return model
+
+
+@MODELS.register_module
+def prmoe_4b(**kwargs):
+    if 'use_residual' in kwargs:
+        kwargs.pop('use_residual')
+    model_kwargs = dict(num_experts=[32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 64, 64],
+                        use_residual=True,
+                        dim=1024,
+                        depth=24,
+                        num_heads=16,
+                        **kwargs)
+    return _create_moegpt_model(**model_kwargs)
+
+
+@MODELS.register_module
+def prmoe_31b(**kwargs):
+    if 'use_residual' in kwargs:
+        kwargs.pop('use_residual')
+    model_kwargs = dict(num_experts=[64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 128, 128],
+                        use_residual=True,
+                        dim=2048,
+                        depth=24,
+                        num_heads=16,
+                        **kwargs)
+    return _create_moegpt_model(**model_kwargs)
+
+
+@MODELS.register_module
+def prmoe_51b(**kwargs):
+    if 'use_residual' in kwargs:
+        kwargs.pop('use_residual')

```

Comment:
use_residual = kwargs.pop('use_residual', False) ?

Reply:
In PR-MOE, 'use_residual' keeps True, but I want key word 'use_residual' removed in dict in case that
it has two same key word in below constructor function.

Reply:
I add warning when there is no 'use_residual' or it's False.